tests/src/JIT/Performance/CodeQuality/Span/Indexer.cs

```diff
@@ -0,0 +1,925 @@
+        public static void SameIndex1(int length)
+        {
+            byte[] a = GetData(length);
+
+            Invoke((int innerIterationCount) =>
+            {
+                Span<byte> s = new Span<byte>(a);
+                byte result = 0;
+                for (int i = 0; i < innerIterationCount; ++i)
+                {
+                    result = TestSameIndex1(s);
+                }
+                return result;
+            },
+            "SameIndex1({0})", length);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static byte TestSameIndex1(Span<byte> data)
+        {
+            byte x = 0;
+            byte y = 0;
+
+            for (var idx = 0; idx < data.Length; idx++)
+            {
+                x ^= data[idx];
+                y ^= data[idx];
+            }
+
+            byte t = (byte)(y ^ x ^ y);
+
+            return t;
+        }
+
+        [Benchmark(InnerIterationCount = Iterations)]
+        [InlineData(DefaultLength)]
+        [Category("Same index in-loop redundant bounds check elimination")]
+        public static void SameIndex2(int length)
+        {
+            byte[] a = GetData(length);
+
+            Invoke((int innerIterationCount) =>
+            {
+                Span<byte> s = new Span<byte>(a);
+                byte result = 0;
+                for (int i = 0; i < innerIterationCount; ++i)
+                {
+                    result = TestSameIndex2(s, ref s[0]);
+                }
+                return result;
+            },
+            "SameIndex2({0})", length);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static byte TestSameIndex2(Span<byte> data, ref byte b)
+        {
+            byte x = 0;
+            byte y = 0;
+            byte ye = 121;
+            byte q = data[0];
+
+            for (var idx = 0; idx < data.Length; idx++)
+            {
+                // Bounds check is redundant, but values are not CSEs.
+                x ^= data[idx];
+                b = 1;
+                y ^= data[idx];
+            }
+
+            byte t = (byte)(y ^ x ^ ye);
+            data[0] = q;
+
+            return t;
+        }
+
+        [Benchmark(InnerIterationCount = Iterations)]
+        [InlineData(DefaultLength)]
+        [Category("Covered index in-loop redundant bounds check elimination")]
+        public static void CoveredIndex1(int length)
+        {
+            byte[] a = GetData(length);
+
+            Invoke((int innerIterationCount) =>
+            {
+                Span<byte> s = new Span<byte>(a);
+                byte result = 0;
+                for (int i = 0; i < innerIterationCount; ++i)
+                {
+                    result = TestCoveredIndex1(s);
+                }
+                return result;
+            },
+            "CoveredIndex1({0})", length);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static byte TestCoveredIndex1(Span<byte> data)
```

Comment:
Our assertion prop currently isn't clever enough to get the cases here, even for arrays.  The covered-index cases we currently handle are:

  - a[K1] followed by a[K2], with K2 >= 0 and K1 >= K2  (with both K1 and K2 constant)
  - a[*] followed by a[0]  (where the "*" means "anything")

Might want to add testing for those (or I can with #10050).